feat: add qs codemod

43081j opened the pull request:
Adds a codemod for migrating from `qs` to `picoquery`.

Some options qs offers are not available. Currently, this mod will not warn on those but should in future.

Draft until we decide these things:
- should it throw when unknown/unsupported options are used?
- when `encode: false` is set, should we replace with `decodeURIComponent(stringify(obj))`? (`encode` is a pretty bad option to have so ideally we should just move consumers off it)
- there are some rarely used options like object sort functions etc that we shouldn't really try codemod. do we just throw when they're used?

thepassle commented:
I think throwing with a clear error message is fine 👍

PuruVJ commented:
This is awkward 😅 https://github.com/es-tooling/module-replacements-codemods/pull/83

thepassle commented:
James and I had a long discussion about this on the e18e discord, and we concluded that migrating `qs` to `pq` might be the better long term solution for projects. While `neoqs` is a great drop-in replacement, `pq` also aligns closer to the URL spec, a lot of work went into the perfomance of it (that's not to say that's not the case for `neoqs`), and `pq` also fixes some of the "correctness" or weirdness/quirks of `qs`. We can discuss more about this on the discord in the [`#cleanup`](https://discord.com/channels/1227627367204257852/1227653370945474590) channel if you like, but let's keep the comments of this PR related to the actual PR 🙂

PuruVJ commented:
No worries, if that's the direction we're going with!

thepassle commented:
> should it throw when unknown/unsupported options are used?

I think it's fine if we bail on these cases/if we can't codemod it

> when encode: false is set, should we replace with decodeURIComponent(stringify(obj))? (encode is a pretty bad option to have so ideally we should just move consumers off it)

Hm that would be a breaking change in behavior, I'm ok with moving consumers off of it, but maybe we should add some logging to let them know about it?

> there are some rarely used options like object sort functions etc that we shouldn't really try codemod. do we just throw when they're used?

I think thats fine yeah

43081j commented:
> Hm that would be a breaking change in behavior, I'm ok with moving consumers off of it, but maybe we should add some logging to let them know about it?

i think that would be sensible

the option leads to all sorts of bugs and weird behaviours so we should 100% be moving people off it. ill see if i can get it to wrap with `decodeURIComponent` then log a warning

43081j commented:
@thepassle i've done the following:

- reworked the options to be a map of option to `Replacer` to make things easier
- added warnings for options we're unaware of (which will be dropped)
- added a warning for `encode: false` and wrapped the result in `decodeURIComponent` automatically
